### PR TITLE
add proxy-mode:ipvs in service.md

### PR DESCRIPTION
**Pull request by Lion-Wei**
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/5571)

### CODE REVIEW

**Comment by Lion-Wei** [comment]
cc @m1093782566 , I added some description about ipvs mode, maybe you can help me review this, thanks

File: docs/concepts/services-networking/service.md [hunk 1]
```diff
@@ -203,6 +203,21 @@ having working [readiness probes](/docs/tasks/configure-pod-container/configure-
 
 ![Services overview diagram for iptables proxy](/images/docs/services-iptables-overview.svg)
 
+### Proxy-mode: ipvs
+
+In this mode, kube-proxy watches the Kubernetes master for the addition and
+removal of `Service` and `Endpoints` objects. For each `Service` it installs
+iptables rules which direct requests for TCP/UDP based services to one of the
```
Comment (m1093782566) [inline docs/concepts/services-networking/service.md]: s/iptables/ipvs

**Comment by m1093782566** [comment]
@Lion-Wei 

Please ref: https://github.com/kubernetes/community/pull/692, you can do better.

**Comment by Lion-Wei** [comment]
@m1093782566  Thanks, I thought now it's better.

File: docs/concepts/services-networking/service.md [hunk 2]
```diff
@@ -203,6 +203,35 @@ having working [readiness probes](/docs/tasks/configure-pod-container/configure-
 
 ![Services overview diagram for iptables proxy](/images/docs/services-iptables-overview.svg)
 
+### Proxy-mode: ipvs
+
+In this mode, kube-proxy watches Kubernetes `services` and `endpoints`,
+call `netlink` interface create ipvs `virtual server` and `real server` accordingly.
```
Comment (tengqm) [inline docs/concepts/services-networking/service.md]: and it creates IPVS "virtual server" and "real server" accordingly.

- a reader doesn't care the fact that `netlink` interface exists and gets called;
- `virtual server` and `real server` are not supposed to be verbatim.

File: docs/concepts/services-networking/service.md [hunk 3]
```diff
@@ -203,6 +203,35 @@ having working [readiness probes](/docs/tasks/configure-pod-container/configure-
...
+Kube-proxy will also sync local virtual server and real server with Kubernetes
+`services` and `endpoints`  periodically, to make sure ipvs status is
+consistent with the expectation. When access the `service`, traffic will
+be redirect to one of the backend `pod`.
+
+Ipvs based on netfilter hook function, like iptables proxy, but use hash
+table as the underlying data structure and work in the kernal state.
+That means ipvs redirects traffic can be much faster, and have much
+better performance when sync proxy rules. Furthermore, ipvs provides more
+options for load balancing algorithm, such as:
```
Comment (tengqm) [inline docs/concepts/services-networking/service.md]: Please fix the grammar errors.

File: docs/concepts/services-networking/service.md [hunk 4]
```diff
@@ -203,6 +203,35 @@ having working [readiness probes](/docs/tasks/configure-pod-container/configure-
...
+### Proxy-mode: ipvs
+
+In this mode, kube-proxy watches Kubernetes `services` and `endpoints`,
+call `netlink` interface create ipvs `virtual server` and `real server` accordingly.
+Kube-proxy will also sync local virtual server and real server with Kubernetes
+`services` and `endpoints`  periodically, to make sure ipvs status is
+consistent with the expectation. When access the `service`, traffic will
+be redirect to one of the backend `pod`.
+
+Ipvs based on netfilter hook function, like iptables proxy, but use hash
```
Comment (m1093782566) [inline docs/concepts/services-networking/service.md]: Similar to iptables, IPVS is based on netfilter...

File: docs/concepts/services-networking/service.md [hunk 5]
```diff
@@ -203,6 +203,35 @@ having working [readiness probes](/docs/tasks/configure-pod-container/configure-
...
+- lc: least connection
+- dh: destination hashing
+- sh: source hashing
+- sed: shortest expected delay
+- nq: never queue
+
+**Note:** ipvs mode assumed IPVS kernel modules are installed on the node
+before running kube-proxy. When kube-proxy starts, if proxy mode is ipvs,
+kube-proxy would validate if IPVS modules are installed on the node, if
+it's not installed kube-proxy will fall back to userspace proxy mode.
```
Comment (m1093782566) [inline docs/concepts/services-networking/service.md]: userspace -> iptables

File: docs/concepts/services-networking/service.md [hunk 6]
```diff
@@ -203,6 +203,35 @@ having working [readiness probes](/docs/tasks/configure-pod-container/configure-
 
 ![Services overview diagram for iptables proxy](/images/docs/services-iptables-overview.svg)
 
+### Proxy-mode: ipvs
```
Comment (caseydavenport) [inline docs/concepts/services-networking/service.md]: This should say `[alpha]` or have some sort of note in bold letters at the top that this is an alpha feature and not recommended for production clusters yet.

**Comment by chenopis** [comment]
FYI, you can use the call formatting for the note: https://kubernetes.io/docs/home/contribute/style-guide/#callout-formatting

**Comment by m1093782566** [comment]
> I think this needs to note that this is an alpha feature, and list any limitations.

+1

**Comment by steveperry-53** [comment]
Ping @Lion-Wei. It looks like this is very close to being ready.

**Comment by zhangxiaoyu-zidif** [comment]
ping @Lion-Wei

**Comment by Lion-Wei** [comment]
@m1093782566 @chenopis @caseydavenport , thank you all, sorry did't fix this in time.

File: docs/concepts/services-networking/service.md (hunk 6 above)
Comment (m1093782566) [reply docs/concepts/services-networking/service.md]: +1

File: docs/concepts/services-networking/service.md (hunk 5 above)
Comment (carlory) [reply docs/concepts/services-networking/service.md]: ipvs.CanUseIPVSProxier will return (false, error) if ipvs modules are not installed. so kube-proxy will fall back to userspace proxy mode. 

I am not sure if I am missing something. @m1093782566 @dujun1990 

```
func tryIPVSProxy(iptver iptables.IPTablesVersioner, kcompat iptables.KernelCompatTester) string {
	// guaranteed false on error, error only necessary for debugging
	// IPVS Proxier relies on iptables
	useIPVSProxy, err := ipvs.CanUseIPVSProxier()
	if err != nil {
		utilruntime.HandleError(fmt.Errorf("can't determine whether to use ipvs proxy, using userspace proxier: %v", err))
		return proxyModeUserspace
	}
	if useIPVSProxy {
		return proxyModeIPVS
	}

	// TODO: Check ipvs version

	// Try to fallback to iptables before falling back to userspace
	glog.V(1).Infof("Can't use ipvs proxier, trying iptables proxier")
	return tryIPTablesProxy(iptver, kcompat)
}
```

Comment (m1093782566) [reply docs/concepts/services-networking/service.md]: Sorry, that's not intended - I assume it's a bug. See bug fix in #54120. FYI

**Comment by steveperry-53** [comment]
@Lion-Wei @m1093782566 @tengqm @carlory @caseydavenport Is this ready to merge? If so, can one of you give tech approval? Thanks.

**Pull request merged**